feat: Add steps APIs to the testmonitor client

SSSantosh18 opened the pull request:
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nisystemlink-clients-python/blob/master/CONTRIBUTING.md).

TODO: Check the above box with an 'x' indicating you've read and followed [CONTRIBUTING.md](https://github.com/ni/nisystemlink-clients-python/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

Include methods for `nitestmonitor/v2/steps/` endpoints in testmonitor client

### Why should this Pull Request be merged?

Steps APIs would be frequently used to extract measurement data from SystemLink for usage in custom scripts or notebook. So a Py client for that would make the script developer's life easy

### What testing has been done?

Manual tested using a Py script

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/_test_monitor_client.py:
```diff
@@ -40,3 +42,164 @@ def api_info(self) -> models.ApiInfo:
+    ) -> Optional[models.UpdateStepsPartialSuccess]:
+        """Updates one or more steps.
+
+        Update requires that the version field matches the version being updated from.
```
Does steps have a version field?

SSSantosh18 replied:
No, it doesn't

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/models/_create_steps_partial_success.py:
```diff
@@ -0,0 +1,25 @@
+
+
+class CreateStepsPartialSuccess(JsonModel):
+    steps: Optional[List[Step]] = None
```
This is not an optional field

SSSantosh18 replied:
it will be empty. made required

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/models/_delete_steps_partial_success.py:
```diff
@@ -0,0 +1,25 @@
+class DeleteStepsPartialSuccess(JsonModel):
+    """The result of deleting multiple steps when one or more steps could not be deleted."""
+
+    steps: Optional[List[StepIdResultIdPair]] = None
```
In swagger this is not given as optional, if there are no steps to delete, will it provide the key and empty list, or do we need to keep it optional?

SSSantosh18 replied:
it will be empty. Made required

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/models/_update_steps_partial_success.py:
```diff
@@ -0,0 +1,25 @@
+
+
+class UpdateStepsPartialSuccess(JsonModel):
+    steps: Optional[List[Step]] = None
```
same here

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/models/_update_steps_request.py:
```diff
@@ -0,0 +1,28 @@
+)
+
+
+class UpdateStepRequestObject(StepRequestObjectBase):
+    name: Optional[str] = None
+    """Step name."""
+
+    children: Optional[List["UpdateStepRequestObject"]] = None
+    """Nested child steps."""
```
Create step request and update step request are both the same, do we need seperate models for this, else we can use the same

SSSantosh18 commented:
> We have to include the api reference updates and example updates apart from the tests. If that is due to the dependency with the results PR, I will wait for the update for the next review

1. Added example
2. Added integration tests, Ideally result has to be created and used for step APIs, for now I have hardcoded the result ID. Will update this once Results API is merged
3. Yet to api reference updates

@SSadaiyappan-NI

SSSantosh18 commented on tests/integration/testmonitor/test_testmonitor.py:
```diff
@@ -9,9 +32,344 @@ def client(enterprise_config: HttpConfiguration) -> TestMonitorClient:
+            text="This is a test step", parameters=[{"name": "param1", "value": "10"}]
+        )
+        properties = {"property1": "value1", "property2": "value2"}
+        # createResultResponse: CreateResultsPartialSuccess = create_result()
```
eventually this create_result() fixture will be used to create result in all tests

priyadarshini-ni replied:
The `create_results` fixture itself has been used as the functionality for both were same

SSadaiyappan-NI commented on examples/testmonitor/steps.py:
```diff
@@ -0,0 +1,93 @@
+)
+
+create_response = client.create_steps(CreateStepsRequest(steps=step_requests))
+# delete steps one by one
+created_steps = create_response.steps
```
```suggestion
created_steps = create_response.steps
# delete steps one by one
```

SSadaiyappan-NI reviewed (changes requested):
Will wait for other changes in the PR

priyadarshini-ni commented on nisystemlink/clients/testmonitor/models/_create_steps_partial_success.py:
```diff
@@ -0,0 +1,25 @@
+    """The list of steps that were successfully created."""
+
+    failed: Optional[List[CreateStepRequest]] = None
+    """The list of steps that were not created.
```
Modified to use `list` in all places

priyadarshini-ni commented on nisystemlink/clients/testmonitor/models/_result.py:
```diff
@@ -42,7 +42,7 @@ class Result(JsonModel):
     """A list of keywords that categorize this result."""
 
-    properties: Optional[Dict[str, str]]
+    properties: Optional[Dict[str, Optional[str]]]
```
The change for this has been done in this [PR](https://github.com/ni/nisystemlink-clients-python/pull/99). This change has been added here for the pipeline check

priyadarshini-ni commented on tests/integration/testmonitor/test_testmonitor.py:
```diff
@@ -396,3 +436,362 @@ def __map_result_to_update_result_request(
+            text="This is a test step", parameters=[{"name": "param1", "value": "10"}]
+        )
+        properties = {"property1": "value1", "property2": "value2"}
+        step = CreateStepRequest(
```
Added

priyadarshini-ni commented on examples/testmonitor/steps.py:
```diff
@@ -0,0 +1,94 @@
+client = TestMonitorClient(configuration=server_configuration)
+
+# Create the step requests
+result_id = "4bef8c3c-afbf-400a-9d30-065dfb0b8d24"
```
Added an example where the result is created and its Id is used to create steps

rbell517 commented on nisystemlink/clients/testmonitor/models/_step.py:
```diff
@@ -0,0 +1,80 @@
+    text: Optional[str] = None
+    """Text string describing the output data."""
+
+    parameters: Optional[List[dict[str, Optional[str]]]] = None
```
I'd like to do better with the steps data than the anything goes the API allows.  We have a standard set of fields to represent certain common parts of a measurement that would go here and the client should encourage using those. What I recommend is a new `Measurement` class with all the standard fields along with an `extras` dictionary for other stuff. 

Here is what I would say is the ideal form.
```python
class Measurement(BaseModel):
    name: Optional[str] = None
    status: Optional[Status] = None
    measurement: Optional[Union[str, int, float]] = None
    lowLimit: Optional[Union[str, int, float]] = None
    highLimit: Optional[Union[str, int, float]] = None
    units: Optional[str] = None
    comparisonType: Optional[str] = None
    extra: Dict[str, str] = {}
```

I say this is the ideal because I'm willing to compromise on this if it makes sense to, for example typing each field as string like it is on the API. I'm concerned about the performance cost for converting these to numeric types if the caller is pulling down a large number of steps with a large number of measurements. Keeping everything as strings would eliminate that. I also don't expect uplink to handle these unions well since the JSON is always going to be a string so it won't know to convert to a number, and neither would we, frankly, we'd basically be guessing. 

cc @adamarnesen and @SSSantosh18 in case they want to add anything. In particular to know how string or numeric typing would impact the data frame we ultimately want to make.

SSadaiyappan-NI replied:
![image](https://github.com/user-attachments/assets/75319f17-f3f2-4577-9365-b31b64757394)
Hi @rbell517 ,

We have added a config to allow extra fields to be added to the model apart from the ones already mentioned. But one drawback we see here is, we can't validate the types of the extra fields in the current pydantic version (1.10.2) we are using. If we have it as a dictionary like you mentioned, we may have to flatten it before sending it to create steps request and put the extra fields inside the extra after querying the steps. We feel that it may take a hit in the performance when we access the extra fields to flatten it and append it with existing data, if there are lot of measurements with extra fields.

If we just provide `Extra: allow`, we can allow extra fields. 
1. If we need to validate fields, we may need to manually iterate through data of each measurement which might affect performance.
2. One more option is to check out the latest Pydantic version and its impact on upgrading it. It might allow us to set the types for extra fields.

Could you provide your thoughts on this?
FYI - @SSSantosh18

rbell517 replied:
> One more option is to check out the latest Pydantic version and its impact on upgrading it.

Unfortunately upgrading to v2 of pydantic has a number of breaking changes so we need to consider the impact.

> they might be able to automatically identify the datatype and consider them as int, float. 

From googling it looks like pandas ability to infer types is so-so, but casting could be reasonable for someone who knows all their measurement data in this analysis workflow are numbers.

I'm OK with keeping the values as strings. I took a look at what pydantic will do with Unions and I agree it will likely be too slow.

For the extras, we still need to be able to access the extra fields for when the Measurement comes as a response. Again newer pydantic versions would help here, they add a native `model_extra` to BaseModel that would surface the extras, but we don't have that. Can you tell me whether extra fields are accessible with what you've implemented? e.g. if the measurement had a `spec_id` field could you access it as `measurement.spec_id`? If so I think that is acceptable.

SSadaiyappan-NI replied:
yes @rbell517 , we are able to access the extra fields just like the defined ones

rbell517 commented on tests/integration/testmonitor/test_testmonitor.py:
```diff
@@ -396,3 +439,417 @@ def __map_result_to_update_result_request(
+            UpdateMultipleStepsRequest(
+                steps=[
+                    UpdateStepRequest(
+                        step_id=step.step_id, result_id=step.result_id, name=new_name
```
Same as in the example, it would be better to update data and inputs/outputs. That would also verify we are transfering those more complex fields properly.

SSadaiyappan-NI replied:
updated as per suggestion

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/models/_step.py:
```diff
@@ -0,0 +1,80 @@
+    """The value."""
+
+
+class StepData(JsonModel):
```
created two seperate files for these models

SSadaiyappan-NI commented on examples/testmonitor/steps.py:
```diff
@@ -0,0 +1,115 @@
+            UpdateStepRequest(
+                step_id=step.step_id,
+                result_id=step.result_id,
+                name="updated name",
```
updated the data instead of the step name

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/_test_monitor_client.py:
```diff
@@ -197,3 +197,165 @@ def delete_results(
+        "results/{resultId}/steps/{stepId}",
+        args=[Path("resultId"), Path("stepId"), Query("updateResultTotalTime")],
+    )
+    def delete_step(
```
updated the order

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/models/_step.py:
```diff
@@ -0,0 +1,80 @@
+    text: Optional[str] = None
+    """Text string describing the output data."""
+
+    parameters: Optional[List[dict[str, Optional[str]]]] = None
```
Updated

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/models/_query_steps_request.py:
```diff
@@ -0,0 +1,141 @@
+
+
+class QueryStepValuesRequest(QueryStepsBase):
+    field: Optional[StepField] = None
```
updated it to be required

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/models/_query_steps_request.py:
```diff
@@ -0,0 +1,141 @@
+    take: Optional[int] = None
+    """Maximum number of steps to return in the current API response."""
+
+    continuation_token: Optional[str] = None
```
updated

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/models/_query_steps_request.py:
```diff
@@ -0,0 +1,141 @@
+    """An enumeration of response formats for step queries."""
+
+    JSON = "JSON"
+    CSV = "CSV"
```
removed the response format

SSadaiyappan-NI commented on nisystemlink/clients/testmonitor/models/_create_steps_request.py:
```diff
@@ -0,0 +1,66 @@
+    """Nested child steps."""
+
+
+class CreateMultipleStepsRequest(JsonModel):
```
updated

rbell517 commented on nisystemlink/clients/testmonitor/models/_step_data.py:
```diff
@@ -0,0 +1,32 @@
+    def __init__(self, **data: Any) -> None:
+        # Convert all extra fields to str while keeping known fields unchanged
+        processed_data = {
+            k: str(v) if k not in self.__annotations__ else v for k, v in data.items()
```
I think you want self.__fields__ instead of __annotations

SSadaiyappan-NI replied:
both will work in our use case, since they both return known properties from the model. I have modified to use __fields__

rbell517 commented on examples/testmonitor/steps.py:
```diff
@@ -0,0 +1,125 @@
+                        lowLimit="30",
+                        highLimit="40",
+                        units="C",
+                        comparisonType="Numeric",
```
Whatever we end up doing with the extra properties we should make sure its shown in the examples. If we need a separate example for that then we should do that. In general I prefer more bite-sized examples that just show how one function works a certain way rather than a chain of examples that build on each other like you have, but both can work.

SSadaiyappan-NI replied:
updated the examples

rbell517 commented on tests/integration/testmonitor/test_testmonitor.py:
```diff
@@ -396,3 +435,461 @@ def __map_result_to_update_result_request(
+
+        assert update_response is not None
+        assert len(update_response.steps) == 1
+        assert update_response.steps[0].data == updated_data
```
The data measurements are special and I would feel better if we broke out that assert into verifying each field to make sure the round trip serialization is working as expected. I'd rather not rely on the pydantic type comparison since we've got extras that I bet it ignores for equality.

SSadaiyappan-NI replied:
Created assert statements for individual fields and the extra fields as well

rbell517 commented on tests/integration/testmonitor/test_testmonitor.py:
```diff
@@ -396,3 +435,461 @@ def __map_result_to_update_result_request(
+                    units="V",
+                    comparisonType="GELE",
+                    specId="spec_01",
+                    specInfo={"specKey": "specValue"},
```
These are only string-string dictionaries, so I'm not sure what this child object will do with your implementation. I guess covert that into a little json snippet?

SSadaiyappan-NI replied:
Based on the implementation, we are converting everything to string. 
```
specInfo="{'specKey': 'specValue'}"
```

rbell517 commented on tests/integration/testmonitor/test_testmonitor.py:
```diff
@@ -396,3 +435,461 @@ def __map_result_to_update_result_request(
+        assert created_step.inputs == inputs
+        assert not created_step.outputs
+
+    def test__create_multiple_steps__multiple_creation_succeed(
```
Would you do a version of this using the `children` field to make a multiple. You'd want to verify the parent is set properly on the children in the response since the response will be flat.

SSadaiyappan-NI replied:
added a new test

rbell517 commented on examples/testmonitor/steps.py:
```diff
@@ -20,7 +19,7 @@ def create_test_result():
             program_name="Example Name",
             host_name="Example Host",
-            status=Status.PASSED(),
+            status="Passed",
```
Results should using the Status class. Measurements just have a string status, but results and steps have the more complex object.

SSadaiyappan-NI replied:
Updated

rbell517 commented on test.py:
Revert this test file before you merge to master

SSadaiyappan-NI replied:
Reverted the file. Did not notice it got committed. Thanks
